src/ConnectionPool.jl

```diff
@@ -459,16 +476,45 @@ function keepalive!(tcp)
     return
 end
 
+struct ConnectTimeout <: Exception
+    host
+    port
+end
+
 function getconnection(::Type{Sockets.TCP},
                        host::AbstractString,
                        port::AbstractString;
                        keepalive::Bool=false,
+                       connect_timeout::Int=0,
                        kw...)::Sockets.TCP
+
     p::UInt = isempty(port) ? UInt(80) : parse(UInt, port)
+
     @debug 2 "TCP connect: $host:$p..."
-    tcp = Sockets.connect(Sockets.getaddrinfo(host), p)
-    keepalive && keepalive!(tcp)
-    return tcp
+
+    if connect_timeout == 0
+        tcp = Sockets.connect(Sockets.getaddrinfo(host), p)
+        keepalive && keepalive!(tcp)
+        return tcp
+    end
+
+    # FIXME this currently causes the Julia runtime to hang on exit...
+    result = Ref{TCPSocket}()
+    error = Ref{Any}()
+    @schedule try
```

Comment:
@vtjnash  I'd like your opinion on this.
Should we include a best-attempt connection timeout feature. Or should we just say for now that underlying infrastructure currently makes this hard to implement, and that users should do their own @async timeout thing if they really need a connect timeout?
https://github.com/JuliaCloud/AWSCore.jl/issues/24#issuecomment-369815567

Reply:
It seems reasonable to add, I just recommended using `close` to do it "properly" (I don't know why the libuv developers claimed in the linked issue that you couldn't cause ECANCELED, since I clearly can):
```julia
julia> s = Sockets.TCPSocket()
Main.Sockets.TCPSocket(RawFD(0xffffffff) init, 0 bytes waiting)

julia> Sockets.connect!(s, Sockets.ip"192.168.101.1", 80)

julia> @async @show Sockets.wait_connected(s)

julia> close(s)

julia>
ERROR (unhandled task failure): connect: operation canceled (ECANCELED)
```

Around this, you just need to add a bit of logic to coordinate closing the socket if the timeout was reached and the socket is still `s.state == StatusConnecting`. And possible also to loop until the timeout is reached (as mentioned in the linked issue, the kernel has its own idea of what the default timeout should be), if you think 2 minutes is too short.

This also uncovers a minor bug in Base, since this assert is wrong:
```diff
diff --git a/stdlib/Sockets/src/Sockets.jl b/stdlib/Sockets/src/Sockets.jl
index ec7876a917..e947da8fd4 100644
--- a/stdlib/Sockets/src/Sockets.jl
+++ b/stdlib/Sockets/src/Sockets.jl
@@ -371,12 +371,13 @@ end
 function uv_connectcb(conn::Ptr{Cvoid}, status::Cint)
     hand = ccall(:jl_uv_connect_handle, Ptr{Cvoid}, (Ptr{Cvoid},), conn)
     sock = @handle_as hand LibuvStream
-    @assert sock.status == StatusConnecting
     if status >= 0
-        sock.status = StatusOpen
+        if !(sock.status == StatusClosed || sock.status == StatusClosing)
+            sock.status = StatusOpen
+        end
         notify(sock.connectnotify)
     else
-        sock.status = StatusInit
+        ccall(:jl_forceclose_uv, Cvoid, (Ptr{Cvoid},), hand)
         err = UVError("connect", status)
         notify_error(sock.connectnotify, err)
     end
```

Reply:
See new commit.
When I tried to use just `close(tcp)` the close call didn't return. If I did `@schedule close(tcp)` the REPL hung on exit. So, for now I'm calling `ccall(:jl_forceclose_uv`, which seems to work.

Is is worth making a PR for stdlib/Socket for this? (It seems like it might be easier to make it robust without being confined to the official public interface) or should it stay in HTTP.jl for now?